Lib/fontTools/subset/svg.py

```diff
@@ -0,0 +1,235 @@
+            if href.startswith("#"):
+                ref_id = href[1:]
+                assert ref_id
+                yield ref_id
+
+        attrs = el.attrib
+        if "style" in attrs:
+            attrs = {**attrs, **parse_css_declarations(el.attrib["style"])}
+        for attr in ("fill", "clip-path"):
+            if attr in attrs:
+                value = attrs[attr]
+                if value.startswith("url(#") and value.endswith(")"):
+                    ref_id = value[5:-1]
+                    assert ref_id
+                    yield ref_id
+
+
+def closure_element_ids(
+    elements: Dict[str, etree.Element], element_ids: Set[str]
+) -> None:
+    # Expand the initial subset of element ids to include ids that can be reached
+    # via references from the initial set.
+    unvisited = element_ids
+    while unvisited:
+        referenced: Set[str] = set()
+        for el_id in unvisited:
+            if el_id not in elements:
+                # ignore dangling reference; not our job to validate svg
+                continue
+            referenced.update(iter_referenced_ids(elements[el_id]))
+        referenced -= element_ids
+        element_ids.update(referenced)
+        unvisited = referenced
+
+
+def subset_elements(el: etree.Element, ids: Set[str]) -> bool:
+    # Keep elements if their id is in the subset, or any of their children's id is.
+    # Drop elements whose id is not in the subset, and either have no children,
+    # or all their children are being dropped.
+    el_id = el.attrib.get("id")
+    if el_id is not None and el_id in ids:
+        # if id is in the set, don't recurse; keep whole subtree
+        return True
+    keep = False
+    for e in el:
+        keep |= subset_elements(e, ids)
+    if keep:
+        return True
+    assert len(el) == 0
+    parent = el.getparent()
+    if parent is not None:
+        parent.remove(el)
+    return False
+
+
+def remap_glyph_ids(
+    elements: Dict[str, etree.Element], glyph_index_map: Dict[int, int]
+) -> Dict[str, str]:
+    # Given {old_gid: new_gid} map, rename all elements containing id="glyph{gid}"
+    # special attributes
+    id_map = {}
+    for el_id, el in elements.items():
+        m = GID_RE.match(el_id)
+        if not m:
+            continue
+        old_index = int(m.group(1))
+        new_index = glyph_index_map.get(old_index)
+        if new_index is not None:
+            if old_index == new_index:
+                continue
+            new_id = f"glyph{new_index}"
+        else:
+            # If the old index is missing, the element correspond to a glyph that was
+            # excluded from the font's subset.
+            # We rename it to avoid clashes with the new GIDs or other element ids.
+            new_id = f".{el_id}"
+            n = count(1)
+            while new_id in elements:
+                new_id = f"{new_id}.{next(n)}"
+
+        id_map[el_id] = new_id
+        el.attrib["id"] = new_id
```

Comment:
This feels like it's doing two jobs and might be simpler if it was broken into handling each explicitly:

1. Assign new gids to things we're keeping
1. Rename things we are dropping to avoid clashes
   * Can we just destroy them? Why do we have to keep the things we aren't keeping and rename them to avoid id clashes?

Reply:
we can't destroy them, because they may contain other elements with id="glyph{digit}" to keep.
I think renaming all the id="glyph{digit}" attributes can logically be done in one go within the same method, whether we are keeping the glyph elements because their id is in our font subset (in which case we want to remap them from old to new glyph indices), or whether we are keeping them indirectly either because they are x-referenced from elements that we are keeping, or because they contain elements that we are keeping.
If we didn't do it in one go, there could be an intermediate state where clashes may still occur (we are updating the tree in-place).
We also return a id_map that has all the final old:new glyph id strings, with the values renamed to match the subset font's new glyph indices or renamed to avoid element id clashes. We can use this in the following `update_glyph_href_links` method where we search for all the `xlink:href="#glyph` references and update those as well.